packages/auth/src/stateMachine/machine.ts

```diff
@@ -1,95 +1,207 @@
 // SPDX-License-Identifier: Apache-2.0
 
+import { Logger } from '@aws-amplify/core';
 import { MachineState } from './MachineState';
+import { v4 as uuid } from 'uuid';
 import {
 	MachineContext,
 	MachineEventPayload,
 	MachineEvent,
 	StateMachineParams,
 	StateTransition,
+	CurrentStateAndContext,
 } from './types';
+import { StateMachineHubEventName } from '../constants/StateMachineHubEventName';
+// TODO: Import from core once library build is resolved
+import { HubClass } from '@aws-amplify/core/lib/Hub';
 
-// TODO: Queue
-// TODO: Listeners
 export class Machine<ContextType extends MachineContext> {
-	name: string;
-	states: Map<string, MachineState<ContextType, MachineEventPayload>>;
-	context: ContextType;
-	current: MachineState<ContextType, MachineEventPayload>;
+	private _name: string;
+	private _states: Map<string, MachineState<ContextType, MachineEventPayload>>;
+	private _context: ContextType;
+	private _current: MachineState<ContextType, MachineEventPayload>;
+	private _logger: Logger;
+	public hub: HubClass;
+	public hubChannel: string;
+	private _eventChannel: string;
+	private _queue: MachineEvent<MachineEventPayload>[];
+	private _queueIdle: boolean = true;
+
 	constructor(params: StateMachineParams<ContextType>) {
-		this.name = params.name;
-		this.states = this._createStateMap(params.states);
-		this.context = params.context;
-		this.current = this.states.get(params.initial) || this.states[0];
+		this._name = params.name;
+		this._states = this._createStateMap(params.states);
+		this._context = params.context;
+		this._queue = [];
+		this._current = this._states.get(params.initial) || this._states[0];
+		this.hub = new HubClass('auth-state-machine');
+		this.hubChannel = `${this._name}-channel`;
+		this._eventChannel = `${this._name}-event-channel`;
+
+		this._logger = new Logger(this._name);
+
+		this.hub.listen(this._eventChannel, hubEvent => {
+			this._queue.push(hubEvent.payload.data.event);
+			if (this._queueIdle) {
+				this._queueIdle = false;
+				this._queueProcessor();
+			}
+		});
 	}
 
 	/**
-	 * Receives an event for processing
+	 * Receives an event for immediate processing
 	 *
 	 * @typeParam PayloadType - The type of payload received in current state
 	 * @param event - The dispatched Event
 	 */
-	send<PayloadType extends MachineEventPayload>(
+	async send<PayloadType extends MachineEventPayload>(
 		event: MachineEvent<PayloadType>
 	) {
-		const validTransition = this.current.findTransition(event);
+		event.id = uuid();
+		this.hub.dispatch(this._eventChannel, {
+			event: StateMachineHubEventName.AUTH_EVENT_ADDED,
+			data: { event },
+		});
+	}
 
-		///TODO: Communicate null transition
-		if (!validTransition) return;
-		const checkGuards = this._checkGuards(validTransition, event);
-		//TODO: Communicate guard failure
-		if (!checkGuards) return;
+	/**
+	 * Returns the current state after previously queued events have been flushed
```

Comment:
nit: `flushed` may not be the most precise word. Normally flush means getting rid of the previous events. I don't think it's the case here. Maybe this is better:
```suggestion
	 * Returns the current state after the state machine has completed reacting to previously queued events
```